File: Go/mergesort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"time"
	"strconv"
	"log"
)

func merge(arr []int, left, mid, right int) {
	n1 := mid - left + 1
	n2 := right - mid
	L := make([]int, n1)
	R := make([]int, n2)
	copy(L, arr[left:mid+1])
	copy(R, arr[mid+1:right+1])
	i, j, k := 0, 0, left
	for i < n1 && j < n2 {
		if L[i] <= R[j] {
			arr[k] = L[i]
			i++
		} else {
			arr[k] = R[j]
			j++
		}
		k++
	}
	for i < n1 {
		arr[k] = L[i]
		i++
		k++
	}
	for j < n2 {
		arr[k] = R[j]
		j++
		k++
	}
}

func mergeSort(arr []int, left, right int) {
	if left < right {
		mid := left + (right-left)/2
		mergeSort(arr, left, mid)
		mergeSort(arr, mid+1, right)
		merge(arr, left, mid, right)
	}
}

func readNumbersFromFile(filePath string) ([]int, error) {
	var numbers []int
	file, err := os.Open(filePath)
	if err != nil {	return nil, err }
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if num, err := strconv.Atoi(line); err == nil {	numbers = append(numbers, num); }
	}
	if err := scanner.Err(); err != nil { return nil, err }
	return numbers, nil
}

func getMemoryUsage() uint64 {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)
	return memStats.Alloc / 1024 // Retorna em KB
}

func getPeakMemoryUsage() uint64 {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)
	return memStats.TotalAlloc / 1024 // Retorna em KB
}

func main() {
	filePath := "arq-desafio.txt"
	initialMemory := getMemoryUsage()
	fmt.Printf("Mem inicial: %d KB\n", initialMemory)

	startTime := time.Now()

	numbers, err := readNumbersFromFile(filePath)
	if err != nil {	log.Fatal("Erro ao ler arq.txt: ", err) }

	afterArrayGenerationMemory := getMemoryUsage()
	fmt.Printf("Mem usada dps de gerar array: %d KB\n", afterArrayGenerationMemory)

	mergeSort(numbers, 0, len(numbers)-1)

	afterSortingMemory := getMemoryUsage()
	fmt.Printf("Mem usada dps de ord: %d KB\n", afterSortingMemory)

	outFile, err := os.Create("arq-saida.txt")
	if err != nil {	log.Fatal("Error ao criar arq-saida.txt: ", err) }
	defer outFile.Close()

	for _, item := range numbers {
		_, err := fmt.Fprintf(outFile, "%d\n", item)
		if err != nil { log.Fatal("Erro ao escrever arq-saida.txt:", err) }
	}

	executionTime := time.Since(startTime).Seconds()
	peakMemory := getPeakMemoryUsage()
	fmt.Printf("Pico de mem usada: %d KB\n", peakMemory)
	fmt.Printf("Tempo de Exec: %.4f seconds\n", executionTime)
}

File: PHP/mergesort.php
<?php

function merge(&$arr, $left, $mid, $right) {
    // Tamanhos dos subarrays
    $n1 = $mid - $left + 1;
    $n2 = $right - $mid;

    // Arrays temporários
    $L = array_slice($arr, $left, $n1);
    $R = array_slice($arr, $mid + 1, $n2);

    // Índices para percorrer os arrays temporários
    $i = 0;  // Índice do array L
    $j = 0;  // Índice do array R
    $k = $left; // Índice do array original

    // Mescla os arrays L e R
    while ($i < $n1 && $j < $n2) {
        if ($L[$i] <= $R[$j]) {
            $arr[$k] = $L[$i];
            $i++;
        } else {
            $arr[$k] = $R[$j];
            $j++;
        }
        $k++;
    }

    // Copia os elementos restantes de L, se houver
    while ($i < $n1) {
        $arr[$k] = $L[$i];
        $i++;
        $k++;
    }

    // Copia os elementos restantes de R, se houver
    while ($j < $n2) {
        $arr[$k] = $R[$j];
        $j++;
        $k++;
    }
}

function mergeSort(&$arr, $left, $right) {
    if ($left < $right) {
        // Encontra o ponto médio
        $mid = floor(($left + $right) / 2);

        // Ordena as duas metades
        mergeSort($arr, $left, $mid);
        mergeSort($arr, $mid + 1, $right);

        // Mescla as duas metades ordenadas
        merge($arr, $left, $mid, $right);
    }
}

function readNumbersFromFile($filePath) {
    $numbers = [];
    $file = fopen($filePath, "r");
    if ($file) {
        while (($line = fgets($file)) !== false) {
            $line = trim($line);
            if (is_numeric($line)) {
                $numbers[] = (int)$line;
            } else {
                echo "Skipping invalid number: $line\n";
            }
        }
        fclose($file);
    } else {
        echo "Error opening the file.\n";
    }

    return $numbers;
}

function getMemoryUsage() {
    return memory_get_usage(true) / 1024; // KB
}

function getPeakMemoryUsage() {
    return memory_get_peak_usage(true) / 1024; // KB
}

$filePath = "arq-desafio.txt";
//$initialMemory = getMemoryUsage();
//echo "Initial memory usage: " . $initialMemory . " KB\n";

$startTime = microtime(true);

$numbers = readNumbersFromFile($filePath);

//$afterArrayGenerationMemory = getMemoryUsage();
//echo "Memory usage after generating array: " . $afterArrayGenerationMemory . " KB\n";

$inicio = 0;
$fim = sizeof($numbers) - 1;

mergeSort($numbers, $inicio, $fim);

//$afterSortingMemory = getMemoryUsage();
//echo "Memory usage after sorting: " . $afterSortingMemory . " KB\n";


$outFile = fopen("arq-saida.txt", "w");
foreach($numbers as $item){
    fwrite($outFile, $item . PHP_EOL);
}
fclose($outFile);

$executionTime = microtime(true) - $startTime;
$peakMemory = getPeakMemoryUsage();
echo "Execution time: " . number_format($executionTime, 4) . " seconds\n";
echo "Peak memory usage: " . $peakMemory . " KB\n";
?>
